### PR TITLE
Ignore nullability when checking for duplicate constraints

### PR DESCRIPTION
**Pull request by cston**
PEVerify requires distinct constraint types.

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 1]
```diff
@@ -35311,6 +35330,92 @@ static void Main()
             Assert.Equal("A2<System.Object?>!", typeParameters[1].ConstraintTypesNoUseSiteDiagnostics[0].ToTestDisplayString(true));
         }
 
+        [Fact]
+        public void UnannotatedConstraint_Override()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: An additional unit test, not related to the code change.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 2]
```diff
@@ -35311,6 +35330,92 @@ static void Main()
...
+                verifyConstraintTypes("B4.F3", "System.String?", "I<System.String?>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String?>!. See TypeSymbolWithAnnotations.SubstituteType.
+                verifyConstraintTypes("B4.F4", "System.String?", "I<System.String?>"); // PROTOTYPE(NullableReferenceTypes): Should be I<System.String?>!. See TypeSymbolWithAnnotations.SubstituteType.
+            }
+
+            void verifyConstraintTypes(string methodName, params string[] expectedTypes)
+            {
+                var constraintTypes = comp.GetMember<MethodSymbol>(methodName).TypeParameters[0].ConstraintTypesNoUseSiteDiagnostics;
+                AssertEx.Equal(expectedTypes, constraintTypes.SelectAsArray(t => t.ToTestDisplayString(true)));
+            }
+        }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: >} [](start = 8, length = 1)

Could you also add a test for the constraint type unification error? `where T : I<string?>, I<T>` and some variants. 
We should probably ignore nullability there too, and produce diagnostics as today in such scenarios.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: There are `TypeUnification_0*` tests. Are there additional variants that should be covered?

**Comment by jcouv** [comment]
> class C3<T, U> : I<T?>, I<U?> where T : struct where U : class { }

Thanks, type unification is indeed already covered.

Note related to this PR: Does `C3` seem right, here? What types `T` and `U` could cause a unification?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs:35005 in 11986c0. [](commit_id = 11986c0cae93bf1b8a72ea0bfbaa98447328fb7d, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs [hunk 3]
```diff
@@ -35311,6 +35330,92 @@ static void Main()
...
+            {
+                verifyConstraintTypes("B1.F1", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B1.F3", "System.String", "I<System.String>");
+                verifyConstraintTypes("B1.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F1", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F2", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F3", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B2.F4", "System.String?", "I<System.String?>");
+                verifyConstraintTypes("B3.F1", "System.String!", "I<System.String!>");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: From our discussion, it's not obvious what should happen when you substitute a `string!` into `T~`.

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs]: Added question to spec-let.

---
In reply to: [199652696](https://github.com/dotnet/roslyn/pull/28248#discussion_r199652696) [](ancestors = 199652696)

**Comment by cston** [comment]
@dotnet-bot test windows_release_unit64_prtest

**Comment by cston** [comment]
@dotnet/roslyn-compiler please review, thanks.

File: docs/features/nullable-reference-types.md [hunk 4]
```diff
@@ -179,15 +179,14 @@ A warning is reported for inconsistent top-level nullability of constraint types
...
 {
-    static void F1<U>() where U : T?, T? { } // error: duplicate constraint
-    static void F2<U>() where U : I<T?>, I<T?> { } // error: duplicate constraint
-    static void F3<U>() where U : T, T? { } // error?
-    static void F4<U>() where U : I<T>, I<T?> { } // error?
+    static void F1<U>() where U : T, T? { } // error: duplicate constraint
+    static void F2<U>() where U : I<T>, I<T?> { } // error: duplicate constraint
 }
 ```
+_If `List<T>` is defined in an unannotated assembly, with `List<string>` is `T` `string!` or `string~`?_
```
[the remark this answers is not in the document]
Comment (jcouv) [reply docs/features/nullable-reference-types.md]: From discussion with Chuck, we've clarified and expanded this question.
Here's a bunch of scenarios with different nullability annotations on the type parameter and the type argument. The question is what is the type of `var`?
Hopefully, that is clearer.

  | List<T!> | List<T~> | List<T?>
-- | -- | -- | --
T = string! |   | C#7: `List<T> F<T>(T t);`   C#8: `var s = F(“”);` | List<string?>
T = string~ | C#8: `List<T> F<T>(T t);`   C#8: `var s = F(obliviousString);` |   | C#8: `List<T?> F<T>(T? t);`   C#8: `var s = F(obliviousString);`
T = string? | List<string?> | C#7: `List<T> F<T>(T t);`   C#8: `var s = F(nullString);` |

[the remark this answers is not in the document]
Comment (cston) [reply docs/features/nullable-reference-types.md]: Added examples from @jcouv.

---
In reply to: [199958208](https://github.com/dotnet/roslyn/pull/28248#discussion_r199958208) [](ancestors = 199958208)

**Pull request merged**